pkg/fsm/metrics/metrics.go

```diff
@@ -143,3 +172,63 @@ func (m *Metrics) RecordSuspend(obj client.Object, suspend bool) {
 	typedObjectRef := meta.MustTypedObjectRefFromObject(obj, m.scheme)
 	m.sink.RecordSuspend(typedObjectRef.ObjectKey(), typedObjectRef.GroupVersionKind(), suspend)
 }
+
+// RecordProcessingStart records the start time of processing for the given GVK and request.
+// This doesn't record a metric, but the start time is used to calculate the processing duration later.
+func (m *Metrics) RecordProcessingStart(
+	gvk schema.GroupVersionKind,
+	req reconcile.Request,
+	gen int64,
+) error {
+	if m.sink == nil || m.options.IsMetricDisabled(types.AchillesProcessingDuration) {
+		return nil
+	}
+
+	// get the processing start time for the given GVK
+	// NOTE: this does not need to be guarded by a mutex because it's guaranteed to only receive concurrent reads
+	processingStartTimes, ok := m.processingStartTimesByGVK[gvk]
+	if !ok {
+		return fmt.Errorf("no processing start time found for GVK %s, missing a call to metrics.InitializeForGVK()", gvk.String())
+	}
+
+	processingStartTimes.Set(req.Name, req.Namespace, gen, time.Now())
+
+	return nil
+}
+
+// RecordProcessingDuration records the time taken to process an object of a given metadata.generation.
+func (m *Metrics) RecordProcessingDuration(
+	gvk schema.GroupVersionKind,
+	req reconcile.Request,
+	gen int64,
+	success bool,
+) error {
+	if m.sink == nil || m.options.IsMetricDisabled(types.AchillesProcessingDuration) {
+		return nil
+	}
+
+	// get the processing start time for the given GVK
+	processingStartTimes, ok := m.processingStartTimesByGVK[gvk]
+	if !ok {
+		return fmt.Errorf("no processing start time found for GVK %s, missing a call to metrics.InitializeForGVK()", gvk.String())
+	}
+
+	// get the processing start time for the given request
+	startTimes := processingStartTimes.GetRange(req.Name, req.Namespace, gen, success)
+
+	now := time.Now()
+	for _, startTime := range startTimes {
+		duration := now.Sub(startTime)
+		m.sink.RecordProcessingDuration(gvk, duration, success)
+	}
+
+	if success {
+		// if the processing was successful, delete all matched items from the tree to prevent unbounded memory growth
+		processingStartTimes.DeleteRange(req.Name, req.Namespace, gen)
+	} else {
+		// if the processing failed, mark all items of (name, namespace) with generation < observedGeneration as failed to avoid subsequent double counting
+		processingStartTimes.SetRangeFailed(req.Name, req.Namespace, gen)
```

Comment:
can't this also lead to unbounded memory growth if we have lots of failures and we never fix things? I'm imagining a dev environment where we know things are broken and just ignore the issue; I wouldn't want this to affect the health of the controller itself.

I think we may want to consider a cleanup policy here as well

Reply:
🔕 I think that's a fair argument -- the spec updates in particular make the risk acceptable